roles/satellite/tasks/internals/activationkey-hostgroup-definition.yml

```
@@ -4,80 +4,92 @@
+    domain: "{{ test_domain | default(omit) }}"
+    architecture: "{{ test_architecture | default(omit) }}"
+    ptable: "{{ test_partition_table | default(omit) }}"
+    operatingsystem: "{{ test_operatingsystem | default(omit) }}"
+    medium: "{{ test_provision_installation_media | default(omit) }}"
+    content_source: "{{ test_foreman_capsule | default(omit) }}"
+    lifecycle_environment: "{{ test_lifecycleenvironment }}"
+    parent: "{{ test_hostgroup_parent | default(omit) }}"
+    pxe_loader: "{{ test_pxe_loader | default(omit) }}"
     locations: "{{ locations.keys() | list }}"
     parameters:
     - name: "kt_activation_keys"
-      value: "{{ g_test_activationkey }}"
+      value: "{{ test_activationkey }}"
     content_view: "{{ deploymentcontentview }}"
-  when: g_test_lifecycleenvironment_defined and g_test_activationkey_defined and g_test_hostgroup_defined
+  when:
+    - test_lifecycleenvironment is defined
+    - deploymentcontentview is defined
+    - test_activationkey is defined
+    - test_hostgroup is defined
 
 - name: "Define Activation Key for production"
   theforeman.foreman.activation_key:
     username: "{{ satellite_user }}"
     password: "{{ satellite_pwd }}"
     server_url: "{{ satellite_url }}"
     validate_certs: "{{ satellite_verify_ssl }}"
-    name: "{{ g_prod_activationkey }}"
+    name: "{{ prod_activationkey }}"
     content_view: "{{ deploymentcontentview }}"
-    lifecycle_environment: "{{ g_prod_lifecycleenvironment }}"
+    lifecycle_environment: "{{ prod_lifecycleenvironment }}"
     subscriptions: "{{ subscriptions | default(omit) }}"
     content_overrides: "{{ contentoverrides | default(omit) }}"
     auto_attach: True
     organization: "{{ satellite_organization }}"
-  when: g_prod_lifecycleenvironment_defined and g_prod_activationkey_defined
+  when:
+    - prod_lifecycleenvironment is defined
+    - deploymentcontentview is defined and
```

I think failing when deploymentcontentview is not defined is fine...

What you are doing is omitting the creation of the activationkey if the variable is not defined, but the whole purpose of this role is to create the activation key, so silently ignoring this misconfiguration is not a good idea.

The deployment contentivew is not optional, while you can choose to have only a production lifecycleenvironment or only a test lifecycle environment.

---

The conditions for creating the following resources have been updated:
- activation key
- lifecycle environment
- host group

Changes:
-  activation key:
Neither lifecycle_environment nor content_view are strictly required (although it's best to have them
The content_view field is now optional.
Since the Library environment always exists, lifecycle_environment now defaults to Library if not configured.

-  host group:
The lifecycle_environment now defaults to Library.
The content_view is already set by satellite_hostgroup_parameters_defaults.

- lifecycle environement:
Removed the check for content_view, as the variable is not even used in its definition.

---

The check on content_view for lifecycle environment is on purpose, as it makes no sense to create a lifecycle environment without putting a content_view inside it.
The only reason to have a lifecycle environment is to put "a number" of contentviews inside. We decided it to be 1:1 to simplify the understanding and cleanup.

I would prefer the activation key to explicitly define a lifecycle environment to avoid misunderstandings. Also using Library is a bad practice so having that by default is not great.
Having an activation key without content_view is weird since the only purpose of the activation key is to assign repositories to the host, so without the content_view it cannot do it.

I don't understand what you are saying about `satellite_hostgroup_parameters_defaults` setting the content_view... this shouldn't be possible since the field `parameters` of a hostgroup is only for ansible/puppet/kickstart and the content_view is neither of them.